chapter-06/inprogress/suggestions.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import "sort"
+
+// bookSuggestions are the suggested books for a given book.
+// Suggestions are the neighbour-books on the shelves.
+type bookSuggestions map[Book]bookCollection
+
+// bookCollection is a collection of books.
+type bookCollection map[Book]struct{}
+
+// suggestOtherBooks returns the suggestions book from the matching-book.
+func suggestOtherBooks(bookworms []Bookworm) []Bookworm {
+	sb := make(bookSuggestions)
+
+	// Register all books on everyone's shelf.
+	for _, bookworm := range bookworms {
+		for i, book := range bookworm.Books {
+			// Each book on the shelf will be a suggestion for those who have read any book on the shelf.
+			otherBooksOnShelves := listOtherBooksOnShelves(i, bookworm.Books)
+			registerBookSuggestions(sb, book, otherBooksOnShelves)
+		}
+	}
+
+	// Suggest a list of related books to each bookworm.
+	suggestions := make([]Bookworm, len(bookworms))
+	for i, bookworm := range bookworms {
+		suggestions[i] = Bookworm{
+			Name:  bookworm.Name,
+			Books: suggestBooks(sb, bookworm.Books),
+		}
+	}
+
+	return suggestions
+}
+
+// listOtherBooksOnShelves returns the list of my books except the one at the given index.
+func listOtherBooksOnShelves(bookIndexToRemove int, myBooks []Book) []Book {
+	// Initialise the first slice: its capacity is the input slice's capacity reduced by 1, and its starting length is the number of items up to the index to discard.
+	otherBooksOnShelves := make([]Book, bookIndexToRemove, len(myBooks)-1)
+	// Copy the slice of books up to the given index into the initialised index.
+	copy(otherBooksOnShelves, myBooks[:bookIndexToRemove])
+	// Append with the rest of myBooks after the index of the discarded book into the created slice.
+	otherBooksOnShelves = append(otherBooksOnShelves, myBooks[bookIndexToRemove+1:]...)
+
+	return otherBooksOnShelves
+}
+
+// registerBookSuggestions registers the books on the same shelf.
+func registerBookSuggestions(suggestions bookSuggestions, reference Book, otherBooksOnShelves []Book) {
+	for _, book := range otherBooksOnShelves {
+		// Check if this reference has already been added to the map.
+		if suggestions[reference] == nil {
```

Comment:
Usually this is written as:

```
collection, ok := suggetions[reference]
if !ok {
    collection = newCollection()
    suggestions[reference] = collection
}

collection.Add(book)
```

There are a few reasons, but mainly it works correctly also for non-nilable values. Similary, it avoids multiple lookups into the map.

Reply:
Fixed but did not implement the `Add` method since we want to avoid methods.... (Start to be annoying to write this code without methods...)